### PR TITLE
Reworked TimeSpanConverter

### PR DESCRIPTION
**Pull request by fazelukario**
- Improved performance of TimeSpanConverter
- Better TimeSpan detection and parsing

### CODE REVIEW

**Comment by OoLunar** [comment]
Have you tested these changes? Could you further explain what you mean by "better parsing"?

**Comment by fazelukario** [comment]
> Have you tested these changes? Could you further explain what you mean by "better parsing"?

Yes I tested it.
"better parsing" Means: Regex string selection improvements (Now the application doesn't need to spend additional resources on removing the d/h/m/s suffix) and TimeSpan is now correctly parsing (For example: The string "25h54m" will be defined as 1 day 1 hour and 54 minutes)

**Comment by OoLunar** [comment]
Oh cool. I'll be doing testing on my side shortly. It's highly likely your PR will be merged once I finish. Thank you very much for your contributions

File: DSharpPlus.CommandsNext/Converters/TimeConverters.cs [hunk 1]
```diff
@@ -56,9 +56,9 @@ public class TimeSpanConverter : IArgumentConverter<TimeSpan>
         static TimeSpanConverter()
         {
 #if NETSTANDARD1_3
-            TimeSpanRegex = new Regex(@"^(?<days>\d+d\s*)?(?<hours>\d{1,2}h\s*)?(?<minutes>\d{1,2}m\s*)?(?<seconds>\d{1,2}s\s*)?$", RegexOptions.ECMAScript);
+            TimeSpanRegex = new Regex(@"^((?<days>\d+)d\s*)?((?<hours>\d+)h\s*)?((?<minutes>\d+)m\s*)?((?<seconds>\d+)s\s*)?$", RegexOptions.ExplicitCapture | RegexOptions.CultureInvariant | RegexOptions.RightToLeft);
```
[the remark this answers is not in the document]
Comment (fazelukario) [reply DSharpPlus.CommandsNext/Converters/TimeConverters.cs:59]: > What's the functional use of `RegexOptions.RightToLeft` in this context?

RightToLeft option used just for regex performance and not affecting output result (it backtracks a lot if done without it).

**Comment by VelvetThePanda** [comment]
I apparently never commented here, but I've written a micro time parser previously, and my implementation could be adapted (mostly by adding the regex flags, but I digress)

```cs
public static class MicroTimeParser
{
    private static readonly Regex _timeRegex = new(@"(?<quantity>\d+)(?<unit>mo|[ywdhms])", RegexOptions.Compiled | RegexOptions.IgnoreCase);

    public static Result<TimeSpan> TryParse(string input)
    {
        var start = TimeSpan.Zero;
        
        var matches = _timeRegex.Matches(input);
        
        if (!matches.Any())
            return Result<TimeSpan>.FromError(new ParsingError<TimeSpan>(input, "Failed to extract time from input."));

        var returnResult = matches.Aggregate(start, (c, n) =>
        {
            var multiplier = int.Parse(n.Groups["quantity"].Value);
            var unit       = n.Groups["unit"].Value;

            return c + unit switch
            {
                "mo" => TimeSpan.FromDays(30  * multiplier),
                "y"  => TimeSpan.FromDays(365 * multiplier),
                "w"  => TimeSpan.FromDays(7   * multiplier),
                "d"  => TimeSpan.FromDays(multiplier),
                "h"  => TimeSpan.FromHours(multiplier),
                "m"  => TimeSpan.FromMinutes(multiplier),
                "s"  => TimeSpan.FromSeconds(multiplier),
                _    => throw new ArgumentOutOfRangeException(nameof(unit), unit, null)
            };
        });
        
        if (returnResult == TimeSpan.Zero)
            return Result<TimeSpan>.FromError(new ParsingError<TimeSpan>(input, "Failed to extract time from input."));
        
        return Result<TimeSpan>.FromSuccess(returnResult);
    }
}
```

As for `Result`, simply `bool` or even `TimeSpan?` could be returned instead

**Comment by OoLunar** [comment]
@DWaffles merged it, not me.

**Comment by DWaffles** [comment]
> @DWaffles merged it, not me.

firefox_4h4d3JpXlz

**Comment by Giggitybyte** [comment]
chrome_Ba1FJGnXpu

**Pull request merged**